CoreCLR runtime tests + Mono on the x64 iOS simulator

imhameed opened the pull request:
This creates another `runtime-staging` lane, named "Build iOSSimulator x64
Release AllSubsets_Mono_RuntimeTests", that will eventually run the runtime
test suite against Mono's non-LLVM JIT on the iOS simulator on amd64 hosts.
Failing tests are added to the exclusion lists in issues.targets.

The tests aren't set to run yet, because they currently take a very long time
to execute.

`AppleAppBuilder` no longer requires a `MainLibraryFileName`. If omitted, one
must be supplied when the app bundle is launched via an environment variable
named `MONO_APPLE_APP_ENTRY_POINT_LIB_NAME`. The generated apps also accept
another environment variable named `MONO_APPLE_APP_ASSEMBLY_LOAD_PREFIX`, which
is a hack used to allow assembly lookup to proceed in a nested app-relative
subdirectory before falling back to the root of the app bundle. This is
necessary because app bundles contain multiple individual test assemblies, and
these assemblies sometimes have dependencies with names that collide with the
dependencies of other test assemblies inside the bundle.

imhameed committed:
Try running the existing runtime test stuff on iOS_x64 and see what breaks

imhameed committed:
Build libraries for iOS x64

imhameed committed:
Checkpoint

imhameed committed:
.

imhameed commented:
Current status: can build apps containing entire runtime test categories. Can also read stdout from instrumented native app launcher code from `xcrun simctl launch --console-pty`. Need to figure out why the tests don't seem to actually run (despite the runtime superficially launching correctly).

imhameed committed:
$ -> @

imhameed commented:
Haven't spent much time on this this week; have mostly been occupied with arm64 simd.

imhameed commented:
I got the tests running in the simulator. I also tried getting these running on-device, but that will require a bit more work, and won't be a part of this PR.

imhameed committed:
Allow specifying a launch assembly via an environment variable

SamMonoRT commented:
No progress made in last few days, will pick up this work next week.

SamMonoRT commented:
@imhameed working on this item this week

imhameed committed:
Add iOSSimulator to the native test component exclusion list

imhameed committed:
iOS -> iOSSimulator

imhameed commented:
baseservices_mono.app/runningmono.dll passed locally:

```
XHarness command issued: apple run --app=baseservices_mono.app --output-directory=out --targets=ios-simulator-64 --mlaunch=/Library/Frameworks/Xamarin.iOS.framework/Versions/14.10.0.4/bin/mlaunch --set-env=MONO_APPLE_APP_ENTRY_POINT_LIB_NAME=runningmono.dll --expected-exit-code=100
info: Preparing run for ios-simulator-64
info: Getting app bundle information..
info: Starting application 'baseservices_mono' on ios-simulator-64
info: App run ended with 100
info: Application has finished with exit code: 100 (as expected)
XHarness exit code: 0
```

imhameed commented:
And xharness now supports app installation and app execution as separate steps (but app uninstalling doesn't yet work in the simulator), after an update to mlaunch: https://github.com/dotnet/xharness/issues/493

```
xharness.sh apple install --app=baseservices_mono.app --output-directory=out --targets=ios-simulator-64 --mlaunch=/Users/imhameed/ms/xharness/artifacts/tmp/Debug/macios-binaries/mlaunch/bin/mlaunch           
XHarness command issued: apple install --app=baseservices_mono.app --output-directory=out --targets=ios-simulator-64 --mlaunch=/Users/imhameed/ms/xharness/artifacts/tmp/Debug/macios-binaries/mlaunch/bin/mlaunch
info: Preparing run for ios-simulator-64
info: Getting app bundle information from '/Users/imhameed/ms/dnr-9.x-ios/runtime/artifacts/tests/coreclr/obj/iOSSimulator.x64.Release/tests/run/iOSApps/baseservices_mono/AppBundle/baseservices_mono/Release-iphonesimulator/baseservices_mono.app'
info: Looking for available ios-simulator-64 simulators..
info: Found simulator device 'iPhone X (iOS 14.4) - created by XHarness'
info: Installing application 'baseservices_mono' on 'iPhone X (iOS 14.4) - created by XHarness'
info: Application 'baseservices_mono' was installed successfully on 'iPhone X (iOS 14.4) - created by XHarness'
XHarness exit code: 0
```
```
xharness.sh apple just-run --app=baseservices_mono.app --output-directory=out --targets=ios-simulator-64 --mlaunch=/Users/imhameed/ms/xharness/artifacts/tmp/Debug/macios-binaries/mlaunch/bin/mlaunch --set-env=MONO_APPLE_APP_ENTRY_POINT_LIB_NAME=runningmono.dll --expected-exit-code=100
XHarness command issued: apple just-run --app=baseservices_mono.app --output-directory=out --targets=ios-simulator-64 --mlaunch=/Users/imhameed/ms/xharness/artifacts/tmp/Debug/macios-binaries/mlaunch/bin/mlaunch --set-env=MONO_APPLE_APP_ENTRY_POINT_LIB_NAME=runningmono.dll --expected-exit-code=100
info: Preparing run for ios-simulator-64
info: Getting app bundle information from '/Users/imhameed/ms/dnr-9.x-ios/runtime/artifacts/tests/coreclr/obj/iOSSimulator.x64.Release/tests/run/iOSApps/baseservices_mono/AppBundle/baseservices_mono/Release-iphonesimulator/baseservices_mono.app'
info: Looking for available ios-simulator-64 simulators..
info: Found simulator device 'iPhone X (iOS 14.4) - created by XHarness'
info: Starting application 'baseservices_mono' on 'iPhone X (iOS 14.4) - created by XHarness'
info: App run ended with 100
info: Application has finished with exit code: 100 (as expected)
XHarness exit code: 0
```

premun commented:
You can disregard that apps don't get uninstalled. It's been like this for years for Xamarin and we also test the whole BCL like that in dotnet/runtime. I don't think the apps get installed in the same way as for real devices and it doesn't cause issues.

If we implement this in the future, it will just start working for you out of the box (if you decide to start calling uninstall now).

SamMonoRT commented on eng/pipelines/runtime.yml:
```diff
@@ -583,6 +583,7 @@ jobs:
     platforms:
```
This is in draft, but just a heads-up - New lanes should be added to the runtime staging yaml

imhameed committed:
Invoke xharness in generated runtime test execution scripts for the iOS simulator

imhameed committed:
aaaaaa

SamMonoRT commented:
@imhameed - I believe you need an entry in ...eng/pipelines/coreclr/templates/helix-queues-setup.yml file to actually kick off the runtime tests. Looking at the iOS simulator results, doesn't seem like it ran the tests.

imhameed committed:
Enable iOSSimulator helix queues

imhameed committed:
...

imhameed committed:
Try dumping output to /tmp/out instead of the home directory, which may require extra permissions to access

imhameed committed:
Build iOS runtime test app bundles in parallel

imhameed committed:
Move test apps to the "XUnitTestBinBase"

premun commented on src/tests/Common/CLRTest.Execute.Bash.targets:
```diff
@@ -365,6 +365,31 @@ fi
+fi
+
+mkdir -p /tmp/out
+$__Command $HARNESS_RUNNER apple run --app="$__TestBinaryBase/$__Category.app" --output-directory="/tmp/out" --set-env="MONO_APPLE_APP_ENTRY_POINT_LIB_NAME=$(MsBuildProjectName).dll" --expected-exit-code=100 -v
```
The problem with running iOS Simulator workloads in Helix is that the way the Helix client is running doesn't work with the Simulator as there's no user GUI session. We work around this by executing the whole payload using `sudo launchctl asuser`. There are couple of more tricks like this (re-launching the `Simulator.app`, detection of Simulator hangs, cleanup..).

Since there are so many caveats and to improve stability, I have tried to capture all this in Helix SDK, specifically in these scripts:
- https://github.com/dotnet/arcade/blob/main/src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/xharness-helix-job.android.sh
- https://github.com/dotnet/arcade/blob/main/src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/xharness-helix-job.apple.sh
- https://github.com/dotnet/arcade/blob/main/src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/xharness-helix-job.apple.sh

Currently, when you define your own command and only ask for the XHarness CLI to be pre-installed, we don't then run your script as part of this wrapper and this will have to be fixed (we have an issue for that that you've seen already https://github.com/dotnet/xharness/issues/590)

premun replied:
With what you try here - you will get timeouts on the simulators (the process spawns but we can't talk to it)

premun replied:
I just got this work prioritized because I see it will be a blocker for you. I estimate around 4 days of work so it would be ready next week. We will then have to come back here and check how you turn your MSBuild items into `HelixWorkItem` so that we can plug it in with what I prepare.

premun commented:
@imhameed can you help me to identify which build and which build leg is this PR creating?

I would like to see the binlog from the build to see how you create the Helix job

SamMonoRT commented:
> @imhameed can you help me to identify which build and which build leg is this PR creating?
> 
> I would like to see the binlog from the build to see how you create the Helix job

It is creating this runtime staging lane : https://github.com/dotnet/runtime/pull/43954/checks?check_run_id=2510681839

premun commented:
I checked how you run Helix and I think the best way to approach this is that in **helixpublishwitharcade.proj**, you will need to create `<XHarnessHelixWorkItem>` instead of `<HelixWorkItem>`. I will then make it include the script inside of the Helix wrapper scripts that will prepare the simulators/devices and clean up after.

Do you share this `.proj` file with others who have non-XHarness workloads? Or you're the only users?

imhameed committed:
Duplicate some script goo

See
https://github.com/dotnet/arcade/blob/01c1af9f5ead4962e390cfbec92396de34118492/src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/xharness-helix-job.apple.sh#L13
and
https://github.com/dotnet/arcade/blob/01c1af9f5ead4962e390cfbec92396de34118492/src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/xharness-runner.apple.sh#L140-L141

imhameed committed:
escaping

imhameed commented:
Weird: the set of tests that failed in the most recent run is not exactly the same as the set of tests that failed in the previous run. Some of the failures are reproducible locally but many are not.

imhameed committed:
Try resetting the simulator (slow) on each test run

imhameed committed:
dummy commit

imhameed committed:
dummy commit

imhameed commented:
Related: https://github.com/dotnet/xharness/issues/596

imhameed committed:
Try using invariant globalization for now (where is icudt.dat supposed to come from?)

imhameed committed:
...

imhameed committed:
Dump the test app's stdout/stderr to stdout. Add a bunch of matching braces to enable use of vi's brace matching feature to quickly find the beginning and end of each log

imhameed committed:
Try not resetting the simulator

imhameed committed:
...

imhameed committed:
Hack: for tests with a "TestGrouping" (see src/tests/Common/testgrouping.proj), unconditionally upload directories with ".app" as a suffix to helix, along with their contents

imhameed committed:
.

imhameed committed:
Hack in simplistic support for loading assemblies from a path relative to the app bundle root

imhameed committed:
Place test assemblies and their supporting files in different subdirectories

imhameed committed:
Prefix test subdirectories

Loader_regressions contains a test DLL named "main", which makes CMake
generate an xcode project that fails to build main.m. I don't really
want to spend time figuring out why CMake is doing this, so instead this
adds a prefix to each test subdirectory.

imhameed committed:
escaping??

imhameed committed:
Add some interop tests to issues.targets

imhameed committed:
Add "PayloadGroup0" reproducible failures to issues.targets

imhameed committed:
Exclude more reproducibly failing tests

imhameed committed:
Exclude another reproducible issue

imhameed committed:
Add more reproducible failures to issues.targets

imhameed committed:
Accumulate more shell script gunk

imhameed committed:
Try OSX.1100.Amd64.Open

imhameed committed:
Revert "Try OSX.1100.Amd64.Open"

This reverts commit 2f0c8cab10e6157bf493a5e11ccb3ce2cdd12aa0.

imhameed committed:
.

imhameed committed:
.

imhameed committed:
Disable another reproducibly failing test

imhameed committed:
Exclude more reproducible failing tests

imhameed committed:
Add more reproducible failures to issues.targets

imhameed committed:
Remove msbuild debug stuff

imhameed commented:
The new lane went green: https://dev.azure.com/dnceng/public/_build/results?buildId=1143523&view=logs&j=4cb2c771-3167-5285-9e21-64bc263d1f6f&t=d48eef33-240c-5153-63be-598e59e13d53

`dci-mac-build-143` took a very long time to run `JIT.Methodical.eE`.

premun commented:
@imhameed can you point me to a log file where you can see the xharness logs? How we call XHarness and which commands?

imhameed commented:
xharness is called by this script: https://github.com/dotnet/runtime/pull/43954/files#diff-3e5a329159ca5b2268e62be8a0d776b6092681e9b241210cb4d57e3454816abcR373-R415

Only `xharness apple run` is used right now. (Android is similar.) Here's the most recent run: https://helix.dot.net/api/jobs/01a12fa7-0660-4bdb-bd7d-d956f19999ee/workitems?api-version=2019-06-17

An example log for `JIT.Methodical.eE`:
https://helix.dot.net/api/jobs/01a12fa7-0660-4bdb-bd7d-d956f19999ee/workitems/JIT.Methodical.eE?api-version=2019-06-17
https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-43954-merge-01a12fa706604bdbbd/JIT.Methodical.eE/console.2a816a29.log?sv=2019-07-07&se=2021-06-07T13%3A32%3A04Z&sr=c&sp=rl&sig=Io09wf8ykr%2BTWHE2jmHNkqPR7k3Obpvzo%2BL%2BaCpyijE%3D

I haven't yet replaced the existing setup with a `CustomCommand` setup that strips away the "XUnitWrapper" and script-invoked-by-XUnitWrapper layers of indirection. I am increasingly coming to believe that it'd make more sense to leave the execution model of this as-is, and then clean up both Android and iOS simultaneously after this has been merged.

SamMonoRT commented on src/tests/issues.targets:
```diff
@@ -3407,4 +3407,164 @@
+
+    <ItemGroup Condition=" $(TargetOS) == 'iOSSimulator' " >
+        <ExcludeList Include = "$(XunitTestBinBase)/Interop/PInvoke/Miscellaneous/MultipleAssembliesWithSamePInvoke/MAWSPITest/**">
+            <Issue>missing assembly</Issue>
```
For this and all below, it is recommended to add GH issue links here. If you haven't yet created these, I would suggest updating this file as a subsequent PR once this is merged.

imhameed replied:
That's the plan.

SamMonoRT replied:
It might make sense to create issues prior to adding new iOS/tvOS device lanes to easily build over common failures.

SamMonoRT commented on src/tests/Common/CLRTest.Execute.Bash.targets:
```diff
@@ -366,6 +366,55 @@ fi
 $__Command $HARNESS_RUNNER android test --instrumentation="net.dot.MonoRunner" --package-name="net.dot.$__Category" --app="$__TestBinaryBase/$__Category.apk" --output-directory=`pwd` --arg=entrypoint:libname=$(MsBuildProjectName).dll --expected-exit-code=100 -v
 CLRTestExitCode=$?
 
+# Exist code of xharness is zero when tests finished successfully
```
nit : Exit code

imhameed committed:
Fix word misuse

SamMonoRT commented on eng/pipelines/coreclr/templates/helix-queues-setup.yml:
```diff
@@ -26,18 +26,22 @@ jobs:
 
+    # iOS/tvOS simulator x64/x86
+    - ${{ if in(parameters.platform, 'iOSSimulator_x64', 'iOSSimulator_x86', 'tvOSSimulator_x64') }}:
+      - OSX.1015.Amd64.Open
```
for my knowledge, will this be used for x86 too ?

imhameed replied:
There are no plans, as far as I know, for an iOS Simulator x86 lane. This was duplicated from `pipelines/libraries/helix-queues-setup.yml`. I can remove the definitions platforms we definitely are not testing as part of https://github.com/dotnet/runtime/issues/33381.

SamMonoRT commented:
@imhameed - Great work. Please outline any follow up work here based on discussions with Premek. Also, please indicate how many tests are actually running, and how many are excluded at the moment.

imhameed committed:
Remove unusued helix pipeline spec for iOSSimulator_x86

SamMonoRT commented:
LGTM - Approving, but need another reviewer to approve prior to merging.

imhameed commented:
The most recent build had a 4-hour timeout expire on the new lane: https://dev.azure.com/dnceng/public/_build/results?buildId=1144258&view=logs&j=4cb2c771-3167-5285-9e21-64bc263d1f6f&t=4cb2c771-3167-5285-9e21-64bc263d1f6f

In this run, the long-running job was `JIT.Methodical`. It ran on `dci-mac-build-146`. All other lanes took much less time to run.

The long-running job from the last test run, `JIT.Methodical.eE`, took [2h44m37s to complete](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-43954-merge-01a12fa706604bdbbd/JIT.Methodical.eE/console.2a816a29.log?sv=2019-07-07&se=2021-06-07T13%3A32%3A04Z&sr=c&sp=rl&sig=Io09wf8ykr%2BTWHE2jmHNkqPR7k3Obpvzo%2BL%2BaCpyijE%3D) on `dci-mac-build-143`. On this run, it took [37m52s to complete](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-43954-merge-5b695068d2a84a9c84/JIT.Methodical.eE/console.1c7a0a2b.log?sv=2019-07-07&se=2021-06-07T20%3A27%3A38Z&sr=c&sp=rl&sig=4Eo5klO2n2mPVOTelh3%2Baj8vdF%2FQpEPWWgHJ7zb%2BSV4%3D) on `dci-macpro-07`.

premun commented:
The queue has MacMinis and MacPros. I imagine MacPros will be way faster, yeah.

premun commented:
@SamMonoRT @imhameed we should definitely not check this in without using the Helix SDK properly (aka `CustomCommands`). I worked with Imran to do the bare minimum to be able to run the simulators in this PR (such as the `launchctl asuser`) but it was never my intention to get this checked in without switching onto the proper usage. The reason is that we do quite a lot of work in the Helix SDK to make sure these workloads run well on Helix and the only way for us to do this by running some code before and after the actual workload.

Not using the Helix SDK properly results in:
- Improper setup and cleanup of the machines. For example it happens about once a week that a Simulator freezes and we resolve that by rebooting the machine and requesting a retry on a different machine. Other than that we kill and prepare the Simulator process so that you don't experience failures.
- Inability to sign apps for device targets (I don't think you care about this at this point, but you might in the future)
- Uploading of `testResults.xml` to AzDO

It's not that you cannot copy the SDK scripts and do similar things here, part of that already happened in this PR just to make it even work. This is more about us needing to change things and deploy them everywhere quite often. For example there is  upcoming work around telemetry where we need to report well for XHarness scenarios which failures are infrastructural and which are caused by the client.

Can we still switch over the to the proper usage before checking in?

fanyang-mono commented on src/tests/Directory.Build.targets:
```diff
@@ -173,7 +173,7 @@
 
      <!-- There are currently no native project binaries on wasm or Android -->
```
Add apple to the comment as well.

fanyang-mono reviewed (approved):
Great work!

SamMonoRT commented:
> @SamMonoRT @imhameed we should definitely not check this in without using the Helix SDK properly (aka `CustomCommands`). I worked with Imran to do the bare minimum to be able to run the simulators in this PR (such as the `launchctl asuser`) but it was never my intention to get this checked in without switching onto the proper usage. The reason is that we do quite a lot of work in the Helix SDK to make sure these workloads run well on Helix and the only way for us to do this by running some code before and after the actual workload.
> 
> Not using the Helix SDK properly results in:
> 
> * Improper setup and cleanup of the machines. For example it happens about once a week that a Simulator freezes and we resolve that by rebooting the machine and requesting a retry on a different machine. Other than that we kill and prepare the Simulator process so that you don't experience failures.
> * Inability to sign apps for device targets (I don't think you care about this at this point, but you might in the future)
> * Uploading of `testResults.xml` to AzDO
> 
> It's not that you cannot copy the SDK scripts and do similar things here, part of that already happened in this PR just to make it even work. This is more about us needing to change things and deploy them everywhere quite often. For example there is upcoming work around telemetry where we need to report well for XHarness scenarios which failures are infrastructural and which are caused by the client.
> 
> Can we still switch over the to the proper usage before checking in?

@imhameed  - we should understand and make the relevant changes. This will help with the device tests which are in the pipeline.

premun commented on src/tests/Common/CLRTest.Execute.Bash.targets:
```diff
@@ -366,7 +366,56 @@ fi
+  --expected-exit-code=100 %5c
+  --targets ios-simulator-64 %5c
+  -v
+CLRTestExitCode=$?
```
Question - the overall Helix job exits with this if at least once you get non-0?

premun commented:
@SamMonoRT @imhameed I did some digging today into the Helix job this new build leg creates and it adds ~16 hours of build time to the `OSX.1015.Amd64.Open` queue (approx 6 times more than the BCL test suite does).

We should definitely not merge this yet because:
- Running this in every `runtime-staging` build will overload the queue which doesn't have the capacity yet - currently, BCL tests already take a good chunk and on busy days it is so-so
- One of the work items here takes almost 4 hours so it needs to be partitioned
- We should do the proper Helix SDK integration

@ilyas1974 @tkapin we should think about how we're going to support these new runtime tests in Helix as we might need way more capacity.
We can figure out some temporary solution too like utilizing one of the Apple device queues which currently get zero traffic.

SamMonoRT commented:
> @SamMonoRT @imhameed I did some digging today into the Helix job this new build leg creates and it adds ~16 hours of build time to the `OSX.1015.Amd64.Open` queue (approx 6 times more than the BCL test suite does).
> 
> We should definitely not merge this yet because:
> 
> * Running this in every `runtime-staging` build will overload the queue which doesn't have the capacity yet - currently, BCL tests already take a good chunk and on busy days it is so-so
> * One of the work items here takes almost 4 hours so it needs to be partitioned
> * We should do the proper Helix SDK integration
> 
> @ilyas1974 @tkapin we should think about how we're going to support these new runtime tests in Helix as we might need way more capacity.
> We can figure out some temporary solution too like utilizing one of the Apple device queues which currently get zero traffic.

Yes, lets iron this out prior to merging : 
1. @imhameed - please update the PR with the customCommands usage and have a green passing lane.
2.  @premun @tkapin : We should discuss a plan to enable these runs. Also note, Imran will be working on trying to add the iOS device lanes soon, also adding an interpreter lane for the simulator as well. We need to make sure there is enough hardware to handle the load, and if there are any tips he should consider while he scripts those up. 
3. cc @marek-safar @steveisok as fyi

akoeplinger commented on src/tests/Common/CLRTest.Execute.Bash.targets:
```diff
@@ -366,7 +366,56 @@ fi
+simulator_app="$xcode_path/Contents/Developer/Applications/Simulator.app"
+open -a "$simulator_app"
+helix_runner_uid="%24(id -u)"
+sudo launchctl asuser "$helix_runner_uid" $__Command $HARNESS_RUNNER apple run %5c
```
For android we introduced separate install/run/uninstall commands so we don't install/uninstall the app a thousand times during the test run which helped improve the time a lot, we probably need the same here: https://github.com/dotnet/xharness/issues/397

premun replied:
We have the same commands available in iOS now: https://github.com/dotnet/xharness/blob/98970876f4d21e6da5198a5c41de51ab1101089f/tests/integration-tests/Apple/Simulator.Commands.Tests.proj#L24-L41

https://github.com/dotnet/xharness/issues/493

akoeplinger replied:
hmm it's a pity that the `android run` command does run-without-install but the `apple run` command does install+run+uninstall and you need `apple just-test`. can we look at coming up with some unified naming in xharness?

premun replied:
Yeah, there's a discrepancy because Apple has two different runtime modes - when it just runs the app and when it does the whole TestRunner listening and whatnot.

I think it would make sense to rename the Android command to `just-test` and have `test` and `just-test` in Android and Apple and then the `run` and `just-run` variants in Apple only?

imhameed replied:
> For android we introduced separate install/run/uninstall commands so we don't install/uninstall the app a thousand times during the test run which helped improve the time a lot, we probably need the same here: [dotnet/xharness#397](https://github.com/dotnet/xharness/issues/397)

I don't think `xharness android install` is being used at all right now by the runtime tests.

fanyang-mono replied:
When https://github.com/dotnet/runtime/pull/54543 gets merged, you should be able to use test only(no more app installation needed here) apple xharness command here.

premun commented:
@SamMonoRT @imhameed we spoke with @ilyas1974 and @tkapin about our options and we think that we first need you to optimize the workloads (for exmaple by not re-installing the applications over and over) to see what kind of build time we're dealing with.

After that we have several options how to proceed such as utilizing other OSX queues (such as the 11.00 one or the AppleTV one) which don't have much traffic at the moment and we will help you target the queues so that we don't overload any single one queue.

fanyang-mono commented:
> @SamMonoRT @imhameed we spoke with @ilyas1974 and @tkapin about our options and we think that we first need you to optimize the workloads (for exmaple by not re-installing the applications over and over) to see what kind of build time we're dealing with.
> 
> After that we have several options how to proceed such as utilizing other OSX queues (such as the 11.00 one or the AppleTV one) which don't have much traffic at the moment and we will help you target the queues so that we don't overload any single one queue.

Just a note here, our team (Sam's team) is looking into enable runtime tests on AppleTV as well. Utilizing AppleTV queue could only be a very temporary solution here. Here is the overall plan of Mono testing. https://github.com/dotnet/runtime/issues/33381 This should help with queue planing.

In terms of the utilizing `android install/run/uninstall`,  I am expecting to work on that with the next week or so.

tkapin commented:
Perfect. We really need to base any utilization estimates on efficient usage of the resources and using the install/test/uninstall commands is supposed to improve the efficiency significantly. Also, it's important to not harm any of the other teams that are using the shared resources.

imhameed committed:
Use just-run instead of run

imhameed committed:
Actually run tests

imhameed committed:
debug app installation failure

imhameed committed:
Hard-code support for 64-bit ios simulators in MobileAppHandler.cs

imhameed committed:
Use app bundle IDs for just-run, rather than the full host paths used for run

fanyang-mono commented:
Tests failed to run and apps failed to uninstall. Currently blocked by https://github.com/dotnet/core-eng/issues/13678

fanyang-mono committed:
add --device

fanyang-mono committed:
Remove unnecessary swith `--device`

imhameed committed:
For now: don't submit tests to helix

fanyang-mono committed:
Adding back the code to make tests run on ios simulator

imhameed committed:
For now: don't submit tests to helix

imhameed commented:
@fanyang-mono @premun I'm going to merge this PR. I've disabled sending the tests to Helix, so this PR won't result in increased use of the OSX.1015.Amd64.Open pool. Followup work (like using CustomCommands, or completing the transition to `xharness apple just-run`, or reducing the end-to-end runtime of a fully-enabled iOSSimulator runtime test lane) will build on the work done in this PR.

akoeplinger reviewed (commented):
LGTM
